docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,391 @@
+		allowed = listsContain(l.whitelistedKeys, l.whitelistedPrefixes, key)
+	}
+	return allowed && !listsContain(l.blacklistedKeys, l.blacklistedPrefixes, key)
+}
+
+func listsContain(keys, prefixes [][]byte, key []byte) bool {
+	for _, k := range keys {
+		if bytes.Equal(key, k) {
+			return true
+		}
+	}
+	for _, p := range prefixes {
+		if bytes.HasPrefix(key, p) {
+			return true
+		}
+	}
+	return false
+}
+```
+
+### ListenKVStore
+We will create a new `Store` type `listenkv.Store` that the `MultiStore` wraps around a `KVStore` to enable state listening.
+This is closely modeled after the `tracekv.Store` with the primary difference being that we can configure the `Store` with a set of `Listening` types
+which direct the streaming of only certain allowed subsets of keys and/or operations to specific `io.Writer` destinations.
+
+```go
+// Store implements the KVStore interface with listening enabled.
+// Operations are traced on each core KVStore call and written to any of the
+// underlying listeners with the proper key and operation permissions
+type Store struct {
+    parent    types.KVStore
+    listeners []types.Listening
+}
+
+// NewStore returns a reference to a new traceKVStore given a parent
+// KVStore implementation and a buffered writer.
+func NewStore(parent types.KVStore, listeners []types.Listening) *Store {
+	return &Store{parent: parent, listeners: listeners}
+}
+
+...
+
+// Get implements the KVStore interface. It traces a read operation and
+// delegates a Get call to the parent KVStore.
+func (tkv *Store) Get(key []byte) []byte {
+	value := tkv.parent.Get(key)
+	writeOperation(tkv.listeners, types.ReadOp, key, value)
+	return value
+}
+
+// Set implements the KVStore interface. It traces a write operation and
+// delegates the Set call to the parent KVStore.
+func (tkv *Store) Set(key []byte, value []byte) {
+	types.AssertValidKey(key)
+	writeOperation(tkv.listeners, types.WriteOp, key, value)
+	tkv.parent.Set(key, value)
+}
+
+// Delete implements the KVStore interface. It traces a write operation and
+// delegates the Delete call to the parent KVStore.
+func (tkv *Store) Delete(key []byte) {
+	writeOperation(tkv.listeners, types.DeleteOp, key, nil)
+	tkv.parent.Delete(key)
+}
+
+// Has implements the KVStore interface. It delegates the Has call to the
+// parent KVStore.
+func (tkv *Store) Has(key []byte) bool {
+	return tkv.parent.Has(key)
+}
+
+...
+
+// writeOperation writes a KVStore operation to the underlying io.Writer of
+// every listener that has permissions to listen to that operation at the given key
+// The TraceOperation is JSON-encoded with the `key` and `value` fields as base64 encoded strings
+func writeOperation(listeners []types.Listening, op types.Operation, key, value []byte) {
+	// short circuit if there are no listeners so we don't waste time base64 encoding `key` and `value`
+	if len(listeners) == 0 {
+		return
+	}
+	traceOp := types.TraceOperation{
```

Comment:
Why do things need to be base64 encoded? Can't we just pass are bytes to listeners?

Reply:
I was following the approach used in the `tracekv` store here, I'll adjust to pass raw bytes instead.